src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+    if (show_time) {
+      uint32_t sec = all_members_time.end_time.tv.tv_sec - all_members_time.start_time.tv.tv_sec;
+      std::cout << "All work-shard start      time = " << all_members_time.start_time << std::endl;
+      std::cout << "All work-shard end        time = " << all_members_time.end_time
+		<< " (" << sec << " seconds)" << std::endl;
+      std::cout << "All work-shard aggregated time = "
+		<< all_members_time.aggregated_time.tv.tv_sec << std::endl;
+    }
+
+    return all_members_time.end_time;
+  }
+
+  //---------------------------------------------------------------------------
+  static void collect_single_shard_stats(const DoutPrefixProvider *dpp,
+					 std::map<std::string, member_time_t> &owner_map,
+					 const shard_progress_t sp_arr[],
+					 unsigned shard,
+					 bool *p_show_time,
+					 const char *name)
+  {
+    const utime_t null_time;
+    const shard_progress_t &sp = sp_arr[shard];
+    if (sp.creation_time == null_time || sp.completion_time == null_time) {
+      *p_show_time = false;
+      return;
+    }
+
+    const std::string &owner = sp.owner;
+    utime_t duration = sp.completion_time - sp.creation_time;
+    if (owner_map.find(owner) != owner_map.end()) {
+      owner_map[owner].aggregated_time += duration;
+      owner_map[owner].end_time = sp.completion_time;
+    }
+    else {
+      owner_map[owner].start_time = sp.creation_time;
+      owner_map[owner].aggregated_time = duration;
+      owner_map[owner].end_time = sp.completion_time;
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::Got " << name
+		       << " stats for shard #" << shard << dendl;
+  }
+
+  //---------------------------------------------------------------------------
+  static void show_dedup_ratio(const worker_stats_t &wrk_stats_sum,
+			       const md5_stats_t    &md5_stats_sum)
+  {
+    uint64_t rados_bytes_before = md5_stats_sum.rados_bytes_before_dedup;
+    uint64_t s3_bytes_before    = wrk_stats_sum.ingress_obj_bytes;
+    uint64_t s3_dedup_bytes     = md5_stats_sum.duplicated_blocks_bytes;
+    uint64_t s3_bytes_after     = s3_bytes_before - s3_dedup_bytes;
+    // skipped objects should be accounted for
+    // TBD: double check the logic!
+    uint64_t skipped_bytes = (wrk_stats_sum.ingress_skip_too_small_bytes +
+			      wrk_stats_sum.non_default_storage_class_objs_bytes);
+    s3_bytes_after -= skipped_bytes;
+    if (rados_bytes_before > s3_bytes_after && s3_bytes_after) {
+      double dedup_ratio = (double)rados_bytes_before/s3_bytes_after;
+      std::cout << "rados_bytes_before = " << rados_bytes_before << "\n";
+      std::cout << "s3_bytes_before    = " << s3_bytes_before << "\n";
+      std::cout << "s3_bytes_after     = " << s3_bytes_after << "\n";
+      std::cout << "dedup_ratio        = " << dedup_ratio << "\n";
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  static void display_progress(uint64_t progress_a, uint64_t progress_b, unsigned shard)
```

Comment:
the plan is to change that to json formatter accumulating the status, right?
`cout` on the completed formatter should happen in the `radosgw-admin` code

Reply:
right